Coinbase updates

robbforce opened the pull request:
My coinbase transactions file also included transfers, which were not handled directly in the v2 parser, so I added them from the transfer parser.
I also added a "conversion" type for coinbase pro.

nanonano commented:
Hi @robbforce, thanks for this, looks like Coinbase have finally fixed their transaction history reports to include transfers, it's only taken them 3 years!

nanonano commented on src/bittytax/conv/parsers/coinbase.py:
```diff
@@ -168,15 +169,31 @@ def _do_parse_coinbase(
+            sell_value=total_ccy,
+            wallet=WALLET,
+        )
+    elif row_dict["Transaction Type"] == "Buy":
```
Trying to understand why you split the handling of "Buy" and "Advanced Trade Buy"?

I don't believe this is related to the new transfers types being added?

Was it because you wanted to add the buy/sell/fee values? If so why are these required, with trades on Coinbase one of the assets is normally fiat already?

robbforce replied:
The buy orders were not converting correctly from my file (missing values) and I didn't have any advanced trades to compare against, so I built the new blocks.

nanonano commented on src/bittytax/conv/parsers/coinbase.py:
```diff
@@ -220,7 +237,27 @@ def _do_parse_coinbase(
             )
-    elif row_dict["Transaction Type"] in ("Sell", "Advanced Trade Sell"):
+    elif row_dict["Transaction Type"] == "Sell":
+        # Saw a sell order for USD to USD? Looks like a bug, so check for these and skip.
```
Noticed this too in my own transaction history report. Looks like Coinbase fixed one thing and broke another.

They are converting any fiat currency to the local currency before a performing a trade, it also means the actual trades are sometimes reported with the wrong currency used.

I can see lots of GBP to GBP trades, also EUR to GBP, but the trade was made with EUR not GBP.

nanonano commented on src/bittytax/conv/parsers/coinbase.py:
```diff
@@ -220,7 +237,27 @@ def _do_parse_coinbase(
                 fee_asset=currency,
                 wallet=WALLET,
             )
-    elif row_dict["Transaction Type"] in ("Sell", "Advanced Trade Sell"):
```
Same question as above, why did you split handling of "Sell" and "Advanced Trade Sell"?

nanonano commented on src/bittytax/conv/parsers/coinbasepro.py:
```diff
@@ -77,17 +77,18 @@ def _parse_coinbase_pro_row(
             wallet=WALLET,
         )
-    elif row_dict["type"] == "match":
+    elif row_dict["type"] in ("conversion", "match"):
```
What is the "conversion" type for? Do you have an example?

robbforce replied:
They look like a buy order and the transactions looked the same as "match" so they were easily handed there.
[conversion_example.csv](https://github.com/BittyTax/BittyTax/files/13795909/conversion_example.csv)

nanonano commented on src/bittytax/conv/parsers/coinbase.py:
```diff
@@ -155,7 +155,9 @@ def _do_parse_coinbase(
             )
     elif row_dict["Transaction Type"] in (
         "Coinbase Earn",
+        "Inflation Reward",
```
What is "Inflation Reward", is it from staking? Can you provide an example.

robbforce replied:
These started showing up once I converted coins to ALGO. At the time, Coinbase would auto stake certain 
coins and pay these rewards.
https://coincodecap.com/coinbase-staking#:~:text=To%20stake%20any%20cryptocurrency%2C%20the,credits%20them%20into%20your%20wallet.
[reward_example.csv](https://github.com/BittyTax/BittyTax/files/13795947/reward_example.csv)

nanonano commented:
Hi @robbforce, I've implemented the transfers part a bit differently, see https://github.com/BittyTax/BittyTax/commit/6bc2c57b22795bec478cdaa2a0a0550d2915c679.

This ensures backwards compatible with previous reports, and balances correctly (if you've only traded using your local currency).

I'm not removing the erroneous fiat to fiat trades, they don't cause any issues, it still balances, and the accounting tool will ignore these. Hopefully Coinbase will fix this bug.

robbforce commented:
> Hi @robbforce, I've implemented the transfers part a bit differently, see [6bc2c57](https://github.com/BittyTax/BittyTax/commit/6bc2c57b22795bec478cdaa2a0a0550d2915c679).
> 
> This ensures backwards compatible with previous reports, and balances correctly (if you've only traded using your local currency).
> 
> I'm not removing the erroneous fiat to fiat trades, they don't cause any issues, it still balances, and the accounting tool will ignore these. Hopefully Coinbase will fix this bug.

I tried this out on my file.
1. The Buy orders fail. Looks like a parse error on the notes: `row[152] ['2021-03-11 06:16:48 UTC', 'Buy', 'USDC', '1000', 'USD', '1.00', '1000.00', '1000.00', '0', 'Bought 1000 USDC for 1000 USD']
ERROR Unexpected Notes content: 'Bought 1000 USDC for 1000 USD'`
2. "Reward Income" and "Inflation Reward" aren't handled yet.

nanonano commented:
> The Buy orders fail. Looks like a parse error on the notes: row[152] ['2021-03-11 06:16:48 UTC', 'Buy', 'USDC', '1000', 'USD', '1.00', '1000.00', '1000.00', '0', 'Bought 1000 USDC for 1000 USD'] ERROR Unexpected Notes content: 'Bought 1000 USDC for 1000 USD'

I fixed this, it was a problem with the regex, it wasn't recognising whole numbers.

`"Reward Income" and "Inflation Reward" aren't handled yet. Note, you're already handling "Rewards Income". Changes found here: https://github.com/BittyTax/BittyTax/pull/332/files/e36aecff841ce1ad37beaa1988e8da8de5355cc3#diff-2399deb443a7c3732ab9edbed2bbc4bdf2a1d434e08b1d56c4e4bd087d2264e9`

I've added both of these as STAKING, instead of just INCOME. Leaving INCOME for those rewards which are linked to learning/watching videos, etc.

Here's the commit https://github.com/BittyTax/BittyTax/commit/b2d18fa67f07df87f642eb860cf0c303e1cd14f9, I've also added the Coinbase Pro change.

Let me know if that works for you. Out of interest, you're in the USA right? What are you using to calculate your US taxes?Hope to support US at some point in the future.

Cheers for all your input. 👍

robbforce commented:
> I've added both of these as STAKING, instead of just INCOME. Leaving INCOME for those rewards which are linked to learning/watching videos, etc.
> 
> Here's the commit [b2d18fa](https://github.com/BittyTax/BittyTax/commit/b2d18fa67f07df87f642eb860cf0c303e1cd14f9), I've also added the Coinbase Pro change.
> 
I left a comment on the commit. My earn rows don't have "Coinbase Earn" as the transaction type.

> Let me know if that works for you. Out of interest, you're in the USA right? What are you using to calculate your US taxes? Hope to support US at some point in the future.
> 
Yeah, I'm in the US. Honestly I struggled to report my crypto taxable events the last two years because I got into mining, defi, staking and have 10k+ transactions, which is cost prohibitive with every online tool available. I've tried ~10 tools and get most transactions into a few of them after writing custom scripts to convert files to their generic formats. None of the tools add up to the same balance. "Accointing" was the closest and has been acquired/retired recently. For actual reporting my taxes I was able to crosscheck with Taxbit the last 2 years, but that tool has gone Enterprise only... Anyway, I'll probably need to hire a CPA for this year's taxes for extra reasons besides the obvious crypto pain, so that's why I'm poking around in Bittytax to at least try to gather and normalize my transactions. I do appreciate all the work and time you've put into this!

nanonano commented:
Thanks you all your help on this, I've pushed a fix for the "Coinbase Earn" as a note here: https://github.com/BittyTax/BittyTax/commit/927ce255039ffaa168d2c37efe92a0ea210371fe.

Seen that a lot, those with large transaction histories are attracted to BittyTax, wishing you all the best getting your taxes sorted.

nanonano commented:
Hi @robbforce, looking for beta testers for the US version of BittyTax :) Details are here: https://github.com/BittyTax/BittyTax/wiki/USA

robbforce commented:
I stopped working on trying to get all my wallets imported because I ended up not selling any crypto last year, so didn't need to report any gains/losses. I also got stuck trying to use Bittytax because I ran into an issue handling multiple files for a single wallet. I was working on adding the staking transactions for Gemini, which come on a separate file with a different header. I added the new file handling and decided to try using the same worksheet name to see if it would combine the transactions. Everything imports with the rows sorting together in a single worksheet fine, which made it easier to see when an asset was staked, but the raw section doesn't combine the columns. It only takes the first files' header and simply places values for all rows in their ordinal position from the original files. It appeared that the way files are stored by wallet in memory would have to be rewritten to handle combining the raw columns discretely.

If you're interested in seeing this directly, I can pull together sample files and push the branch with the Gemini grow changes to Github.

Another item that became a showstopper was that I had some liquidity pool transactions on BSC and saw that multiple inbound assets can't be handled. I haven't thought of a workaround, but have since discovered that cryptotaxcalculator.io manages these. Their ledger appears to be organized simply as transaction type (in/out), asset, amount, value, then they break down the macro transaction into these smaller parts and probably link them by an id or the transaction hash.

Anyway, my situation is complex enough that I'm not able to get to the accounting and tax reporting methods of Bittytax. The starred exchanges/wallets are running cleanly through the converter and I had a lot more to go:

2016
  Ripple
2020
  * Coinbase, added transaction types
  Ember
2021
  * Coinbase pro, added conversion transactions
  * Nexo, added new headers and transaction types
  * Celsius, added transaction types
  Nicehash
  Valora
  - Gemini, needs more transaction types
  * Binance.us, added parser
  * Binance
  Probit
  * Yoroi (ADA)
  * MyAlgo (ALGO)
  Trust (BNB)
  - Trust (BSC), unable to handle multiple inbound assets
  Trust (ETH)
  Trust (LTC)
  * Trust (LUNC), worksheet label is broken with multiple staketax imports (cosmetic)
  Trust (MATIC)
  Trust (RVN)
  Trust (ZIL)
  Metamask (MATIC)
2022
  - Kucoin, use generic parser?
  Firefly (IOTA, SMR)
  Terra (LUNA)
  * Terra (LUNC), worksheet label is broken with multiple staketax imports (cosmetic)
2023
  - Uphold, needs fixes to staking
  Ronin
  Pera (ALGO)
2024
  Pi
________________________________
From: Scott Green ***@***.***>
Sent: Monday, March 18, 2024 1:00 PM
To: BittyTax/BittyTax ***@***.***>
Cc: Nathaniel Roark ***@***.***>; Mention ***@***.***>
Subject: Re: [BittyTax/BittyTax] Coinbase updates (PR #332)


Hi @robbforce<https://github.com/robbforce>, looking for beta testers for the US version of BittyTax :) Details are here: https://github.com/BittyTax/BittyTax/wiki/USA

—
Reply to this email directly, view it on GitHub<https://github.com/BittyTax/BittyTax/pull/332#issuecomment-2004831674>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AADXNH3WV6E45225CLZTDTDYY5BWXAVCNFSM6AAAAABBBGRIQKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMBUHAZTCNRXGQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

nanonano commented:
Hi @robbforce, appreciate your detailed reply.

>  I also got stuck trying to use Bittytax because I ran into an issue handling multiple files for a single wallet.

Please raise an issue for this, and anything else where you found problems, or you think needs improvement.

> I was working on adding the staking transactions for Gemini, which come on a separate file with a different header. I added the new file handling and decided to try using the same worksheet name to see if it would combine the transactions. 

> If you're interested in seeing this directly, I can pull together sample files and push the branch with the Gemini grow changes to Github.

Yes please do a PR for Gemini staking and provide samples. These worksheets won't get combined automatically as they have a different header format, but BittyTax will process and order them correctly.

It would be possible to write a merge parser which would combine the rows, and map the raw data as it best fits. I can look at this.

Interested to know what issues you had with the Uphold parser, are you able to provide a sample export?

How did you get on with your Kucoin transactions?

Thanks again!
